lightning-invoice/src/lib.rs

```diff
@@ -1365,9 +1364,32 @@ pub mod utils {
 			invoice = invoice.route(hint);
 		}
 
-		invoice.build_signed(|msg_hash| {
-			secp_ctx.sign_recoverable(msg_hash, &channelmanager.get_our_node_secret())
-		})
+		let raw_invoice = invoice.build_raw().unwrap();
+		let hrp_str = raw_invoice.hrp.to_string();
+		let hrp_bytes = hrp_str.as_bytes();
+		let data_without_signature = raw_invoice.data.to_base32();
+		let mut invoice_preimage = Vec::<u8>::from(hrp_bytes);
+
+		let mut data_part = Vec::from(data_without_signature);
+		let overhang = (data_part.len() * 5) % 8;
+		if overhang > 0 {
+			// add padding if data does not end at a byte boundary
+			data_part.push(u5::try_from_u8(0).unwrap());
+
+			// if overhang is in (1..3) we need to add u5(0) padding two times
+			if overhang < 3 {
+				data_part.push(u5::try_from_u8(0).unwrap());
+			}
+		}
+
+		invoice_preimage.extend_from_slice(&Vec::<u8>::from_base32(&data_part)
+		  .expect("No padding error may occur due to appended zero above."));
+		let signature = match keys_manager.sign_invoice(invoice_preimage) {
+			Ok(sig) => sig,
+			Err(_) => return Err(SemanticError::InvalidSignature)
+		};
+		let signed_raw_invoice: Result<SignedRawInvoice, SemanticError> = raw_invoice.sign(|_| Ok(signature));
+		Invoice::from_signed(signed_raw_invoice.unwrap())
```

Comment:
The signing API is designed to allow custom errors. So the error type here should be what is returned by `sign_invoice`. Something like:

```rust
let signed_raw_invoice = raw_invoice.sign(|_| keys_manager.sign_invoice(invoice_preimage))?;
```

 But may have to convert to a `SignOrCreationError`.